### ISSUE TITLE
Inference time not optimized (continiue from #1087)

### ISSUE THREAD

**Issue by mpanahish** [neuralmagic/sparseml#1120]
previous issue : #1087 
I'd use below code to transfer learning yolov5 on my custom dataset
`!sparseml.yolov5.train --data {dataset.location}/data.yaml --cfg models_v5.0/yolov5s.yaml \ --hyp data/hyps/hyp.finetune.yaml \ --recipe zoo:cv/detection/yolov5-s/pytorch/ultralytics/coco/pruned-aggressive_96?recipe_type=transfer --batch 16\ --epochs 100 --imgsz 416 --save-period 5`

and after that, export to onnx by `!sparseml.yolov5.export_onnx` command.

I expect this onnx file give faster inference than yolov5s traind on same custom data which exported to onnx.

but i get this result for sparssified yolov5s :
on CPU: 345ms
on GPU: 13 ms

and this for yolov5s :
on CPU: 345 ms
on GPU: 12 ms

**Comment by robertgshaw** [comment]
Which runtime are you using for the CPU result? Are you using DeepSparse Runtime?

**Comment by mpanahish** [comment]
i use onnxruntime(CPU build)	library 

like this  guide : [https://onnxruntime.ai/docs/api/python/api_summary.html#load-and-run-a-model](url)

**Comment by rsnm2** [comment]
ONNX Runtime is not optimized to run sparse models. Neural Magic's DeepSparse is a runtime that is specifically designed to take advantage of kernel sparsity. Because DeepSparse knows the location of the 0's ahead of time, it is able to "skip" the multiply-adds by 0, which dramatically speeds up computation.

As such, to see a big speedup from sparsity, try running with DeepSparse. DeepSparse accepts an ONNX file as input.

You can see an example here: https://docs.neuralmagic.com/use-cases/object-detection/deploying

**Comment by mpanahish** [comment]
> 

i did this way.
but inference time is increased !
average inference time for 100 image on CPU is 497.
last time without DeepSparse was 345

**Comment by mpanahish** [comment]
> 


i did this way.
but inference time is increased !
average inference time for 100 image on CPU is 497.
last time without DeepSparse was 345

**Comment by mpanahish** [comment]
another try :
i test inference time by using sparse YOLOv5 model with my image test.

`from deepsparse import Pipeline

model_stub = "zoo:cv/detection/yolov5-l/pytorch/ultralytics/coco/pruned-aggressive_98"
yolo_pipeline = Pipeline.create(
    task="yolo",
    model_path=model_stub,
)

import pandas as pd
df = pd.DataFrame(columns=('category', 'time' ,'std', 'desc'))
  
for im , image_path in  get_image(10) :
  m = %timeit -o yolo_pipeline(images=[image_path], iou_thres=0.6, conf_thres=0.001)
  i = len(df)
  df.loc[i] = ['inference',m.average , m.stdev , image_path]
`

average inference time for 100 image on CPU is 1.8 s.

**Comment by rsnm2** [comment]
@mpanahish - what type of system are you running on (how many cores)?

Can you try running the deepsparse.benchmark script - this is an automated performance testing script?

deepsparse.benchmark [stub]

**Comment by SunHaozhe** [comment]
> Neural Magic's DeepSparse is a runtime that is specifically designed to take advantage of kernel sparsity.

Does this imply that with this library (SparseML), we will not see any inference speedup when using unstructured pruning (sparsity of atomic parameters/elements/weights instead of kernels/channels)?

**Comment by bfineran** [comment]
Hi @mpanahish running your benchmark in a `Pipeline` will add extra overhead that may not be measured in your other experiments. For example, the Pipeline call includes loading the image, (simple) preprocessing, and postprocessing + NMS.
Calls like loading the image will be a fixed cost that even if they take a few ms will add up in your sum of 100 calls.

Per @rsnm2's suggestion, you can try using the `deepsparse.benchmark` utility with `--engine deepsparse` and then `--engine onnxruntime` to compare performance between those runtimes.  Additionally, any information you can provide on your CPU can help us further optimize your deployment.

**Comment by bfineran** [comment]
> > Neural Magic's DeepSparse is a runtime that is specifically designed to take advantage of kernel sparsity.
> 
> Does this imply that with this library (SparseML), we will not see any inference speedup when using unstructured pruning (sparsity of atomic parameters/elements/weights instead of kernels/channels)?

Hi @SunHaozhe SparseML itself isn't a runtime, but just a framework for sparsifying and exporting these models. Our **DeepSparse repository and engine will see speedups for unstructured pruning**, as well as other sparsification techniques.

"Kernel sparsity" here refers to the entire parameter (kernel) of a NN layer, in other words unstructured.  There is a bit of a name overlap with "kernel" in the context of a convolutional layer - so see how you got here.

**Comment by jeanniefinks** [comment]
Hello @mpanahish and @SunHaozhe 
As we haven't heard anything further on this, we are going to assume you're all set. I will close out this issue but if you would like to continue the conversation, please re-open the issue. Thank you for your questions.
Regards, Jeannie - Neural Magic
